hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/io/hops/yarn/server/resourcemanager/quota/QuotaService.java

```diff
@@ -0,0 +1,454 @@
+          connector.writeLock();
+          computeAndApplyChargeInt(startData.getContainerId(), resourcesMap, usedMillis);
+          connector.commit();
+          return null;
+        }
+
+      };
+      quotaSchedulerHandler.handle();
+      if (!remove) {
+        containerQuotaTime.put(startData.getContainerId(), endTime);
+      }
+    }
+  }
+
+  private void computeAndApplyChargeInt(ContainerId containerId, Map<String, Long> resourcesMap, long usedMillis) throws
+      StorageException {
+    //Get Data  ** ProjectQuota **
+    ProjectQuotaDataAccess<ProjectQuota> pqDA = (ProjectQuotaDataAccess) RMStorageFactory.getDataAccess(ProjectQuotaDataAccess.class);
+    ProjectsDailyCostDataAccess<ProjectDailyCost> pdcDA = (ProjectsDailyCostDataAccess) RMStorageFactory.getDataAccess(
+        ProjectsDailyCostDataAccess.class);
+    
+    final long curentDay = TimeUnit.DAYS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+    // Calculate the quota
+    // Get ApplicationId from ContainerId
+    ApplicationId appId = containerId.getApplicationAttemptId().getApplicationId();
+
+    //Get ProjectId from ApplicationId in ** ApplicationState Table ** 
+    String appOwner;
+
+    RMApp app = rmContext.getRMApps().get(appId);
+    if (app == null) {
+      LOG.error("Application not found: " + appId.toString() + " for container " + containerId);
+      return;
+    } else {
+      appOwner = app.getUser();
+
+    }
+
+
+
+    String projectName = HopsWorksHelper.getProjectName(appOwner);
+    String user = HopsWorksHelper.getUserName(appOwner);
+
+    //Get Data  ** ProjectQuota **
+    ProjectQuota projectQuota = pqDA.get(projectName);
+    ProjectDailyCost projectDailyCost = pdcDA.get(projectName, user, curentDay);
+    float charge = computeCharge(resourcesMap, usedMillis);
+    chargeProjectQuota(projectQuota, projectName, user, containerId, charge);
+    //** ProjectDailyCost charging**
+    chargeProjectDailyCost(projectDailyCost, projectName, user, curentDay, charge, appId);
+    if (projectQuota != null) {
+      pqDA.add(projectQuota);
+    }
+    if(projectDailyCost!=null){
+      pdcDA.add(projectDailyCost);
+    }
+  }
+
+  private float computeCharge(Map<String, Long> resourcesMap, long usedMillis) {
+    //the pricePerTick is set for a minimum sized container, the price to pay is
+    //proportional to the container size on the most used resource
+    float multiplicator = rmContext.getPriceMultiplicatorService().getMultiplicator(PriceMultiplicator.MultiplicatorType.GENERAL);
+    float vcoresUsage = (float) resourcesMap.get(ResourceInformation.VCORES_URI) / minVcores;
+    float vcoresCredit = (float) usedMillis / DateUtils.MILLIS_PER_SECOND * vcoresUsage * multiplicator * basePriceGeneral;
+    float memoryUsage = (float) resourcesMap.get(ResourceInformation.MEMORY_URI) / minMemory;
+    float memoryCredit = (float) usedMillis / DateUtils.MILLIS_PER_SECOND * memoryUsage * multiplicator * basePriceGeneral;
```

Comment:
Divide by zero on all these divisions if *usage is 0? I don't think this could ever happen though

Reply:
The usage is not part of the division, it is:
(usedMillis / DateUtils.MILLIS_PER_SECOND) * vcoresUsage * multiplicator * basePriceGeneral;
I will add a check for the mins that should never be null.